File: src/main/java/dev/goerner/geozen/model/jackson/deserializer/GeometryDeserializer.java
package dev.goerner.geozen.model.jackson.deserializer;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import dev.goerner.geozen.model.Geometry;
import dev.goerner.geozen.model.LineString;
import dev.goerner.geozen.model.MultiLineString;
import dev.goerner.geozen.model.MultiPoint;
import dev.goerner.geozen.model.MultiPolygon;
import dev.goerner.geozen.model.Point;
import dev.goerner.geozen.model.Polygon;

import java.io.IOException;

public class GeometryDeserializer extends AbstractGeometryDeserializer<Geometry> {

	@Override
	public Geometry deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
		JsonNode rootNode = p.getCodec().readTree(p);
		String type = rootNode.get("type").asText();

		JavaType javaType = switch (type) {
			case "Point" -> ctxt.constructType(Point.class);
			case "LineString" -> ctxt.constructType(LineString.class);
			case "Polygon" -> ctxt.constructType(Polygon.class);
			case "MultiPoint" -> ctxt.constructType(MultiPoint.class);
			case "MultiLineString" -> ctxt.constructType(MultiLineString.class);
			case "MultiPolygon" -> ctxt.constructType(MultiPolygon.class);
			default -> throw new IllegalArgumentException("Invalid GeoJSON type: " + type + ".");
		};
		JsonDeserializer<?> deserializer = ctxt.findRootValueDeserializer(javaType);

		return (Geometry) deserializer.deserialize(rootNode.traverse(p.getCodec()), ctxt);
	}
}

File: src/main/kotlin/dev/goerner/geozen/jackson/deserializer/GeometryDeserializer.kt
package dev.goerner.geozen.jackson.deserializer

import dev.goerner.geozen.model.Geometry
import dev.goerner.geozen.model.collections.GeometryCollection
import dev.goerner.geozen.model.multi_geometry.MultiLineString
import dev.goerner.geozen.model.multi_geometry.MultiPoint
import dev.goerner.geozen.model.multi_geometry.MultiPolygon
import dev.goerner.geozen.model.simple_geometry.LineString
import dev.goerner.geozen.model.simple_geometry.Point
import dev.goerner.geozen.model.simple_geometry.Polygon
import tools.jackson.core.JsonParser
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.JsonNode

class GeometryDeserializer : AbstractGeometryDeserializer<Geometry>() {

    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): Geometry {
        val rootNode = p.readValueAsTree<JsonNode>()

        val typeNode = rootNode["type"]
        require(typeNode != null && typeNode.isString) { "Geometry must have a valid 'type' field." }

        val typeClass = when (val type = typeNode.asString()) {
            "Point" -> Point::class.java
            "LineString" -> LineString::class.java
            "Polygon" -> Polygon::class.java
            "MultiPoint" -> MultiPoint::class.java
            "MultiLineString" -> MultiLineString::class.java
            "MultiPolygon" -> MultiPolygon::class.java
            "GeometryCollection" -> GeometryCollection::class.java
            else -> throw IllegalArgumentException("Invalid GeoJSON type: $type.")
        }

        return ctxt.readValue(rootNode.traverse(ctxt), typeClass)
    }
}
